Enable gocritic linter and fix issues

alexandear opened the pull request:
This PR enables `gocritic` linter. Also, fixes or disables `gocritic` issues.

The full log of running `golangci-lint`:
```sh
❯ golangci-lint run
pkg/qemu/entitlementutil/entitlementutil.go:52:14: preferStringWriter: ent.WriteString(entXML) should be preferred to the ent.Write([]byte(entXML)) (gocritic)
        if _, err = ent.Write([]byte(entXML)); err != nil {
                    ^
pkg/qemu/entitlementutil/entitlementutil.go:65:1: whyNoLint: include an explanation for nolint directive (gocritic)
// isColimaWrapper__useThisFunctionOnlyForPrintingHints returns true
^
pkg/executil/command.go:19:18: ptrToRefParam: consider `ctx' to be of non-pointer type (gocritic)
func WithContext(ctx *context.Context) Opt {
                 ^
pkg/networks/usernet/dnshosts/dnshosts_test.go:209:1: paramTypeCombine: func(i int, j int) bool could be replaced with func(i, j int) bool (gocritic)
func (r recordSorter) Less(i int, j int) bool {
^
pkg/networks/usernet/dnshosts/dnshosts_test.go:214:1: paramTypeCombine: func(i int, j int) could be replaced with func(i, j int) (gocritic)
func (r recordSorter) Swap(i int, j int) {
^
pkg/networks/usernet/dnshosts/dnshosts_test.go:226:1: paramTypeCombine: func(i int, j int) bool could be replaced with func(i, j int) bool (gocritic)
func (z zoneSorter) Less(i int, j int) bool {
^
pkg/networks/usernet/dnshosts/dnshosts_test.go:231:1: paramTypeCombine: func(i int, j int) could be replaced with func(i, j int) (gocritic)
func (z zoneSorter) Swap(i int, j int) {
^
pkg/networks/usernet/dnshosts/dnshosts_test.go:197:21: typeUnparen: could simplify (recordSorter) to recordSorter (gocritic)
        _ sort.Interface = (recordSorter)(nil)
                           ^
pkg/networks/usernet/dnshosts/dnshosts_test.go:198:21: typeUnparen: could simplify (zoneSorter) to zoneSorter (gocritic)
        _ sort.Interface = (zoneSorter)(nil)
                           ^
pkg/qemu/imgutil/imgutil.go:82:1: paramTypeCombine: func(source string, dest string) error could be replaced with func(source, dest string) error (gocritic)
func ConvertToRaw(source string, dest string) error {
^
pkg/textutil/textutil.go:28:1: paramTypeCombine: func(prefix string, text string) string could be replaced with func(prefix, text string) string (gocritic)
func PrefixString(prefix string, text string) string {
^
pkg/textutil/textutil.go:47:1: paramTypeCombine: func(cutset string, text string) string could be replaced with func(cutset, text string) string (gocritic)
func TrimString(cutset string, text string) string {
^
pkg/textutil/textutil.go:52:1: paramTypeCombine: func(message string, text string) string could be replaced with func(message, text string) string (gocritic)
func MissingString(message string, text string) string {
^
pkg/sshutil/sshutil.go:225:1: paramTypeCombine: func(instDir string, useDotSSH, forwardAgent bool, forwardX11 bool, forwardX11Trusted bool) ([]string, error) could be replaced with func(instDir string, useDotSSH, forwardAgent, forwardX11, forwardX11Trusted bool) ([]string, error) (gocritic)
func SSHOpts(instDir string, useDotSSH, forwardAgent bool, forwardX11 bool, forwardX11Trusted bool) ([]string, error) {
^
pkg/networks/commands.go:78:1: deprecatedComment: the proper format is `Deprecated: <text>` (gocritic)
// Deprecated. Use Sock.
^
pkg/networks/config.go:178:1: deprecatedComment: the proper format is `Deprecated: <text>` (gocritic)
// Deprecated. Use Sock.
^
pkg/downloader/downloader.go:294:1: paramTypeCombine: func(cacheDir string, remote string) string could be replaced with func(cacheDir, remote string) string (gocritic)
func cacheDirectoryPath(cacheDir string, remote string) string {
^
pkg/downloader/downloader.go:383:1: paramTypeCombine: func(dst, src, ext string, description string) error could be replaced with func(dst, src, ext, description string) error (gocritic)
func decompressLocal(dst, src, ext string, description string) error {
^
pkg/downloader/downloader.go:475:1: paramTypeCombine: func(localPath, url string, description string, expectedDigest digest.Digest) error could be replaced with func(localPath, url, description string, expectedDigest digest.Digest) error (gocritic)
func downloadHTTP(localPath, url string, description string, expectedDigest digest.Digest) error {
^
pkg/limayaml/defaults.go:711:1: paramTypeCombine: func(format string, instDir string) (bytes.Buffer, error) could be replaced with func(format, instDir string) (bytes.Buffer, error) (gocritic)
func executeHostTemplate(format string, instDir string) (bytes.Buffer, error) {
^
pkg/limayaml/defaults_test.go:424:21: appendAssign: append result not assigned to the same slice (gocritic)
        expect.Provision = append(y.Provision, d.Provision...)
                           ^
pkg/limayaml/defaults_test.go:425:18: appendAssign: append result not assigned to the same slice (gocritic)
        expect.Probes = append(y.Probes, d.Probes...)
                        ^
pkg/limayaml/defaults_test.go:426:24: appendAssign: append result not assigned to the same slice (gocritic)
        expect.PortForwards = append(y.PortForwards, d.PortForwards...)
                              ^
pkg/limayaml/defaults_test.go:427:22: appendAssign: append result not assigned to the same slice (gocritic)
        expect.CopyToHost = append(y.CopyToHost, d.CopyToHost...)
                            ^
pkg/limayaml/defaults_test.go:428:31: appendAssign: append result not assigned to the same slice (gocritic)
        expect.Containerd.Archives = append(y.Containerd.Archives, d.Containerd.Archives...)
                                     ^
pkg/limayaml/defaults_test.go:429:27: appendAssign: append result not assigned to the same slice (gocritic)
        expect.AdditionalDisks = append(y.AdditionalDisks, d.AdditionalDisks...)
                                 ^
pkg/limayaml/defaults_test.go:432:18: appendAssign: append result not assigned to the same slice (gocritic)
        expect.Mounts = append(d.Mounts, y.Mounts...)
                        ^
pkg/limayaml/defaults_test.go:433:20: appendAssign: append result not assigned to the same slice (gocritic)
        expect.Networks = append(d.Networks, y.Networks...)
                          ^
pkg/limayaml/defaults_test.go:588:18: appendAssign: append result not assigned to the same slice (gocritic)
        expect.Mounts = append(d.Mounts, y.Mounts...)
                        ^
pkg/store/disk.go:63:1: unnamedResult: consider giving a name to these results (gocritic)
func inspectDisk(fName string) (int64, string, error) {
^
pkg/store/disk.go:82:1: unnamedResult: consider giving a name to these results (gocritic)
func inspectDiskWithQemuImg(fName string) (int64, string, error) {
^
pkg/networks/usernet/config.go:32:1: paramTypeCombine: func(dir string, name string, sockType SockType) (string, error) could be replaced with func(dir, name string, sockType SockType) (string, error) (gocritic)
func SockWithDirectory(dir string, name string, sockType SockType) (string, error) {
^
pkg/networks/usernet/config.go:111:1: paramTypeCombine: func(baseIP net.IP, netMask net.IP) (net.IP, *net.IPNet, error) could be replaced with func(baseIP, netMask net.IP) (net.IP, *net.IPNet, error) (gocritic)
func netmaskToCidr(baseIP net.IP, netMask net.IP) (net.IP, *net.IPNet, error) {
^
pkg/networks/usernet/gvproxy_test.go:36:1: paramTypeCombine: func(t *testing.T, file string, content string) could be replaced with func(t *testing.T, file, content string) (gocritic)
func createResolveFile(t *testing.T, file string, content string) {
^
pkg/networks/usernet/gvproxy_test.go:41:11: preferStringWriter: writer.WriteString(content) should be preferred to the writer.Write([]byte(content)) (gocritic)
        _, err = writer.Write([]byte(content))
                 ^
pkg/cidata/cidata.go:440:3: deferInLoop: Possible resource leak, 'defer' is called in the 'for' loop (gocritic)
                defer f.Close()
                ^
pkg/qemu/qemu.go:160:1: paramTypeCombine: func(cfg Config, cmd string, tag string) (string, error) could be replaced with func(cfg Config, cmd, tag string) (string, error) (gocritic)
func sendHmpCommand(cfg Config, cmd string, tag string) (string, error) {
^
pkg/qemu/qemu.go:316:1: paramTypeCombine: func(exe string, machine string) (*features, error) could be replaced with func(exe, machine string) (*features, error) (gocritic)
func inspectFeatures(exe string, machine string) (*features, error) {
^
pkg/qemu/qemu.go:465:1: unnamedResult: consider giving a name to these results (gocritic)
func Cmdline(cfg Config) (string, []string, error) {
^
pkg/qemu/qemu.go:1018:1: unnamedResult: consider giving a name to these results (gocritic)
func Exe(arch limayaml.Arch) (string, []string, error) {
^
pkg/vz/vm_darwin.go:693:1: unnamedResult: consider giving a name to these results (gocritic)
func createSockPair() (*os.File, *os.File, error) {
^
pkg/hostagent/hostagent.go:668:1: paramTypeCombine: func(ctx context.Context, sshConfig *ssh.SSHConfig, port int, local, remote string, verb string, reverse bool) error could be replaced with func(ctx context.Context, sshConfig *ssh.SSHConfig, port int, local, remote, verb string, reverse bool) error (gocritic)
func forwardSSH(ctx context.Context, sshConfig *ssh.SSHConfig, port int, local, remote string, verb string, reverse bool) error {
^
pkg/hostagent/hostagent.go:735:8: uncheckedInlineErr: removeErr error is unchecked, maybe intended to check it instead of err (gocritic)
                                if removeErr := os.RemoveAll(local); err != nil {
                                   ^
pkg/hostagent/mount.go:50:3: assignOp: replace `sshfsOptions = sshfsOptions + ",cache=no"` with `sshfsOptions += ",cache=no"` (gocritic)
                sshfsOptions = sshfsOptions + ",cache=no"
                ^
pkg/hostagent/mount.go:53:3: assignOp: replace `sshfsOptions = sshfsOptions + ",follow_symlinks"` with `sshfsOptions += ",follow_symlinks"` (gocritic)
                sshfsOptions = sshfsOptions + ",follow_symlinks"
                ^
pkg/hostagent/port.go:45:1: unnamedResult: consider giving a name to these results (gocritic)
func (pf *portForwarder) forwardingAddresses(guest api.IPPort, localUnixIP net.IP) (string, string) {
^
pkg/hostagent/port_darwin.go:19:1: paramTypeCombine: func(ctx context.Context, sshConfig *ssh.SSHConfig, port int, local, remote string, verb string) error could be replaced with func(ctx context.Context, sshConfig *ssh.SSHConfig, port int, local, remote, verb string) error (gocritic)
func forwardTCP(ctx context.Context, sshConfig *ssh.SSHConfig, port int, local, remote string, verb string) error {
^
cmd/limactl/gendoc.go:120:1: paramTypeCombine: func(dir string, old, new string) error could be replaced with func(dir, old, new string) error (gocritic)
func replaceAll(dir string, old, new string) error {
^
cmd/limactl/gendoc.go:136:10: wrapperFunc: use bytes.ReplaceAll method in `bytes.Replace(in, []byte(old), []byte(new), -1)` (gocritic)
                out := bytes.Replace(in, []byte(old), []byte(new), -1)
                       ^
```

AkihiroSuda commented:
CI is failing

alexandear commented:
@AkihiroSuda fixed

AkihiroSuda commented on pkg/guestagent/guestagent_linux.go:
```diff
@@ -47,15 +47,17 @@ func New(newTicker func() (<-chan time.Time, func()), iptablesIdle time.Duration
 	if auditStatus.Enabled == 0 {
-		if err = auditClient.SetEnabled(true, libaudit.WaitForReply); err != nil {
+		err = auditClient.SetEnabled(true, libaudit.WaitForReply)
+		if err != nil {
```
Why?

alexandear replied:
It's the [`sloppyReassign`](https://go-critic.com/overview.html#sloppyreassign) check.

There are two ways of fixing this:

1)
```
if err := auditClient.SetEnabled(true, libaudit.WaitForReply); err != nil {
```

2)
```
err = auditClient.SetEnabled(true, libaudit.WaitForReply); 
if err != nil {
```

I choose the second because it doesn't change semantics of the previous code.

AkihiroSuda replied:
The check seems too strict

alexandear replied:
Disabled `sloppyReassign` check and reverted changes.

alexandear commented on pkg/limayaml/defaults_test.go:
```diff
@@ -585,7 +585,7 @@ func TestFillDefault(t *testing.T) {
 
 	// o.Mounts just makes d.Mounts[0] writable because the Location matches
-	expect.Mounts = append(d.Mounts, y.Mounts...)
+	expect.Mounts = append(append([]Mount{}, d.Mounts...), y.Mounts...)
```
This is the fixing for [`appendAssign`](https://go-critic.com/overview.html#appendassign) check:

```
pkg/limayaml/defaults_test.go:432:18: appendAssign: append result not assigned to the same slice (gocritic)
        expect.Mounts = append(d.Mounts, y.Mounts...)

```

jandubois replied:
I'm not sure I agree with the `sloppyReassign` check, but the `appendAssign` one makes sense to me: to make it explicit that `append` modifies the slices in-place and doesn't return a copy.

alexandear commented on .golangci.yml:
```diff
@@ -18,7 +18,6 @@
 ---
 run:
   concurrency: 6
-  deadline: 5m
```
Removed because the property `deadline` is missing in [v1.55.2](https://github.com/golangci/golangci-lint/blob/e3c2265f4939976874989e159386b3bb7dcf8e1f/.golangci.reference.yml#L8)

alexandear commented on cmd/limactl/gendoc.go:
```diff
@@ -133,7 +133,7 @@ func replaceAll(dir string, old, new string) error {
 		if err != nil {
 			return err
 		}
-		out := bytes.Replace(in, []byte(old), []byte(new), -1)
```
```sh
cmd/limactl/gendoc.go:136:10: wrapperFunc: use bytes.ReplaceAll method in `bytes.Replace(in, []byte(old), []byte(new), -1)` (gocritic)
                out := bytes.Replace(in, []byte(old), []byte(new), -1)
                       ^
```

alexandear commented on pkg/cidata/cidata.go:
```diff
@@ -437,10 +437,11 @@ func writeCIDataDir(rootPath string, layout []iso9660util.Entry) error {
 		if err != nil {
 			return err
 		}
-		defer f.Close()
```
```sh
pkg/cidata/cidata.go:440:3: deferInLoop: Possible resource leak, 'defer' is called in the 'for' loop (gocritic)
                defer f.Close()
                ^
```

alexandear commented on pkg/downloader/downloader.go:
```diff
@@ -291,7 +291,7 @@ func Cached(remote string, opts ...Opt) (*Result, error) {
 // cacheDirectoryPath returns the cache subdirectory path.
 // - "url" file contains the url
 // - "data" file contains the data
-func cacheDirectoryPath(cacheDir string, remote string) string {
```
```sh
pkg/downloader/downloader.go:294:1: paramTypeCombine: func(cacheDir string, remote string) string could be replaced with func(cacheDir, remote string) string (gocritic)
func cacheDirectoryPath(cacheDir string, remote string) string {
^
```

alexandear commented on pkg/executil/command.go:
```diff
@@ -16,6 +16,8 @@ type options struct {
 
 // WithContext runs the command with CommandContext.
+//
+//nolint:gocritic // consider `ctx' to be of non-pointer type
```
```
pkg/executil/command.go:19:18: ptrToRefParam: consider `ctx' to be of non-pointer type (gocritic)
func WithContext(ctx *context.Context) Opt {
                 ^
```

alexandear commented on pkg/hostagent/hostagent.go:
```diff
@@ -732,7 +733,7 @@ func forwardSSH(ctx context.Context, sshConfig *ssh.SSHConfig, port int, local,
 				}
 			} else {
 				logrus.WithError(err).Warnf("Failed to set up forward from %q (guest) to %q (host)", remote, local)
-				if removeErr := os.RemoveAll(local); err != nil {
```
```
pkg/hostagent/hostagent.go:735:8: uncheckedInlineErr: removeErr error is unchecked, maybe intended to check it instead of err (gocritic)
                                if removeErr := os.RemoveAll(local); err != nil {
                                   ^
```

alexandear commented on pkg/hostagent/mount.go:
```diff
@@ -47,10 +47,10 @@ func (a *HostAgent) setupMount(m limayaml.Mount) (*mount, error) {
 	// NOTE: allow_other requires "user_allow_other" in /etc/fuse.conf
 	sshfsOptions := "allow_other"
 	if !*m.SSHFS.Cache {
-		sshfsOptions = sshfsOptions + ",cache=no"
```
```
pkg/hostagent/mount.go:50:3: assignOp: replace `sshfsOptions = sshfsOptions + ",cache=no"` with `sshfsOptions += ",cache=no"` (gocritic)
                sshfsOptions = sshfsOptions + ",cache=no"
                ^
```

alexandear commented on pkg/qemu/entitlementutil/entitlementutil.go:
```diff
@@ -49,7 +49,7 @@ func Sign(qExe string) error {
     <true/>
   </dict>
 </plist>`
-	if _, err = ent.Write([]byte(entXML)); err != nil {
```
```
pkg/qemu/entitlementutil/entitlementutil.go:52:14: preferStringWriter: ent.WriteString(entXML) should be preferred to the ent.Write([]byte(entXML)) (gocritic)
        if _, err = ent.Write([]byte(entXML)); err != nil {
                    ^
```

alexandear commented on pkg/qemu/entitlementutil/entitlementutil.go:
```diff
@@ -68,7 +68,7 @@ func Sign(qExe string) error {
 // The result can be used *ONLY* for controlling hint messages.
 // DO NOT change the behavior of Lima depending on this result.
 //
-//nolint:revive
```
```
pkg/qemu/entitlementutil/entitlementutil.go:65:1: whyNoLint: include an explanation for nolint directive (gocritic)
// isColimaWrapper__useThisFunctionOnlyForPrintingHints returns true
^
```

alexandear commented on pkg/qemu/qemu.go:
```diff
@@ -462,9 +462,9 @@ func qemuMachine(arch limayaml.Arch) string {
 	return "virt"
 }
 
-func Cmdline(cfg Config) (string, []string, error) {
```
```
pkg/qemu/qemu.go:465:1: unnamedResult: consider giving a name to these results (gocritic)
func Cmdline(cfg Config) (string, []string, error) {
^
```

alexandear commented on pkg/limayaml/defaults_test.go:
```diff
@@ -421,16 +421,16 @@ func TestFillDefault(t *testing.T) {
 
 	expect = y
 
-	expect.Provision = append(y.Provision, d.Provision...)
```
```
pkg/limayaml/defaults_test.go:424:21: appendAssign: append result not assigned to the same slice (gocritic)
        expect.Provision = append(y.Provision, d.Provision...)
                           ^
```

alexandear commented on pkg/networks/commands.go:
```diff
@@ -75,7 +75,7 @@ func (config *YAML) Sock(name string) string {
 
 // VDESock returns a vde socket.
 //
-// Deprecated. Use Sock.
```
```
pkg/networks/commands.go:78:1: deprecatedComment: the proper format is `Deprecated: <text>` (gocritic)
// Deprecated. Use Sock.
^
```

alexandear commented:
Disabled `sloppyReassign` test, changed gocritic's configuration to use `enabled-tags` instead of `enabled-checks`, and added comments to lines with issues (marked as "Resolved").

AkihiroSuda reviewed (approved):
Thanks
